### PR TITLE
Fix startup race condition

### LINKED ISSUE #3366: Race condition error: listen EACCES: permission denied ::0:25
I am testing latest master and some(!!) machines has serious problem starting haraka at all.

**previous versions:**
```
2024-05-21 14:30:10.573300500  [NOTICE] [-] [core] Listening on [::0]:25
2024-05-21 14:30:10.573718500  [NOTICE] [-] [server] Switching from current gid: 0
2024-05-21 14:30:10.574419500  [NOTICE] [-] [server] New gid: 8
2024-05-21 14:30:10.574470500  [NOTICE] [-] [server] Switching from current uid: 0
2024-05-21 14:30:10.574761500  [NOTICE] [-] [server] New uid: 88
```

**master:**
```
2024-05-21 13:44:41.261384500  [NOTICE] [-] [server] Switching from current gid: 0
2024-05-21 13:44:41.262136500  [NOTICE] [-] [server] New gid: 8
2024-05-21 13:44:41.262237500  [NOTICE] [-] [server] Switching from current uid: 0
2024-05-21 13:44:41.262567500  [NOTICE] [-] [server] New uid: 88
...
2024-05-21 13:44:41.275302500  [ERROR] [-] [server] Failed to setup listeners: listen EACCES: permission denied ::0:25
2024-05-21 13:44:41.278932500  [ERROR] [-] [server] Error: listen EACCES: permission denied ::0:25
2024-05-21 13:44:41.278940500      at Server.setupListenHandle [as _listen2] (node:net:1800:21)
2024-05-21 13:44:41.278944500      at listenInCluster (node:net:1865:12)
2024-05-21 13:44:41.278948500      at doListen (node:net:2014:7)
2024-05-21 13:44:41.278952500      at process.processTicksAndRejections (node:internal/process/task_queues:83:21) {
2024-05-21 13:44:41.278956500    code: 'EACCES',
2024-05-21 13:44:41.278959500    errno: -13,
2024-05-21 13:44:41.278963500    syscall: 'listen',
2024-05-21 13:44:41.278967500    address: '::0',
2024-05-21 13:44:41.278970500    port: 25
2024-05-21 13:44:41.278974500  }
```

It seems issue is opening port after droping priviledges (missing await?). I am going to investigate it now

### PR DESCRIPTION
**Pull request by analogic**
Fixes #3366

Changes proposed in this pull request:
- making endpoint.bind() to async and awaiting in server.js

Checklist:
- [ ] docs updated
- [x] tests updated
- [x] [Changes](https://github.com/haraka/Haraka/blob/master/Changes.md) updated

### CODE REVIEW

**Comment by msimerson** [comment]
It looks like you also need to await [here](https://github.com/haraka/Haraka/pull/3367/files?diff=unified&w=1#diff-a4c65ede64197e1a112899a68bf994485b889c4b143198bac4af53425b38406fL508)

**Comment by analogic** [comment]
PR tested on multiple instances, seems OK
```
2024-05-21 20:25:39.085531500  [NOTICE] [-] [server] Listening on [::0]:25
2024-05-21 20:25:39.085777500  [NOTICE] [-] [server] Switching from current gid: 0
2024-05-21 20:25:39.086551500  [NOTICE] [-] [server] New gid: 8
2024-05-21 20:25:39.086589500  [NOTICE] [-] [server] Switching from current uid: 0
2024-05-21 20:25:39.087241500  [NOTICE] [-] [server] New uid: 88
```

**Review by msimerson — CHANGES_REQUESTED** [review]
LGTM.

I'd like to see one more change. Now that bind is async, I think we should be using `fs/promises` instead of sync calls:

```diff
diff --git a/endpoint.js b/endpoint.js
index 8154aef1..8f8e959a 100644
--- a/endpoint.js
+++ b/endpoint.js
@@ -1,7 +1,7 @@
 'use strict';
 // Socket address parser/formatter and server binding helper
 
-const fs = require('node:fs');
+const fs = require('node:fs/promises');
 const sockaddr = require('sockaddr');
 
 module.exports = function endpoint (addr, defaultPort) {
```

Since `fs.exists` has been deprecated for years, switch it to use `fs.access` or `fs.stat` before the `await fs.unlink(path);`

**Comment by msimerson** [comment]
or [ignore errors](https://nodejs.org/api/fs.html#fspromisesrmpath-options) when the file doesn't exist:

```js
await fs.rm(this.path, { force: true });
```

File: endpoint.js [hunk 1]
```diff
@@ -46,21 +46,24 @@ class Endpoint {
...
-        else {
+            opts.path = this.path;
+            if (fs.existsSync(opts.path)) fs.unlinkSync(opts.path);
+        } else {
             opts.host = this.host;
             opts.port = this.port;
         }
-        server.listen(opts, done);
+        
+        return await new Promise((resolve, reject) => {
```
Comment (msimerson) [inline endpoint.js:61]:
```suggestion
        return new Promise((resolve, reject) => {
```

There's no need to await here.

**Comment by analogic** [comment]
When I "grep node:fs" I didn't find many "promises" so I assumed it would be better not to change. I chose "rm" for simplicity...

**Review by msimerson — APPROVED** [review]

**Comment by msimerson** [comment]
Thanks! Great job on the diagnosis and the fix.

**Pull request merged**